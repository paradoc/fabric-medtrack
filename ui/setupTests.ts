import '@testing-library/jest-dom/extend-expect'


import dayjs from 'dayjs'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import duration from 'dayjs/plugin/duration'
import relativeTime from 'dayjs/plugin/relativeTime'

dayjs.extend(isSameOrAfter)
dayjs.extend(duration)
dayjs.extend(relativeTime)


export function setupFetchStub(data: any) {
  return function fetchStub(_url: string) {
    return new Promise((resolve) => {
      resolve({
        json: () => Promise.resolve(data),
      })
    })
  }
}