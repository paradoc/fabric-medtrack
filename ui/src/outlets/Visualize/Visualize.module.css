.container {
  display: grid;
  grid-template-rows: max-content minmax(180px, 360px);
  row-gap: 48px;
  place-items: center;
  height: 100%;
  width: 100%;
}

.container > header {
  font-weight: bold;
  font-size: 20px;
  color: #666;
  line-height: 28px;
  text-align: center;
}