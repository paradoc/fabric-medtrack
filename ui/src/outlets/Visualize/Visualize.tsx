import React, { useEffect, useState, useMemo } from 'react'
import { useLocation } from 'react-router'
import { prop } from 'rambda'
import {
  BarChart,
  Bar,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from 'recharts'

import styles from './Visualize.module.css'
import { Link } from 'react-router-dom'

type QueryResult = {
  current: number
  expected: number
  generic_name: string
  total_dispatches: number
}

export default function Visualize() {
  const [data, setData] = useState<QueryResult[]>([])
  const [visData, setVisData] = useState<QueryResult[]>([])
  const [hasQueried, setHasQueried] = useState(false)
  const { search } = useLocation()
  const searchParams = useMemo(() => new URLSearchParams(search), [search])
  const medications = searchParams
    .get('g')
    ?.split(',')
    .map((m) => m.trim())
  const start = searchParams.get('s')
  const end = searchParams.get('e')

  useEffect(() => {
    if (medications && medications.length > 0 && start && end && !hasQueried) {
      Promise.all([
        ...medications.map((med) =>
          fetch(`/api/query?generic=${med}&start=${start}&end=${end}`).then(
            (r) => r.json()
          )
        ),
      ]).then((r) => {
        setData(r)
        setHasQueried(true)
      })
    }
  }, [medications, start, end, hasQueried])

  useEffect(() => {
    if (hasQueried && data.length > 0) {
      const sanitized = data.filter((d) => d.total_dispatches !== 0)
      setVisData(sanitized)
    }
  }, [data, hasQueried])

  return (
    <div className={styles.container}>
      <header>
        Compliance rates for medicines:{' '}
        {visData.map(prop('generic_name')).join(', ')} from {start} to {end}
      </header>
      {visData.length > 0 && (
        <div
          style={{ height: '100%', width: '100%', flex: 1 }}
          data-testid="chart"
        >
          <ResponsiveContainer height="100%" width="100%">
            <BarChart
              width={640}
              height={480}
              data={visData}
              margin={{
                top: 5,
                right: 30,
                left: 20,
                bottom: 5,
              }}
            >
              <CartesianGrid strokeDasharray="3 3" />
              <XAxis dataKey="generic_name" />
              <YAxis />
              <Tooltip />
              <Legend />
              <Bar dataKey="current" fill="#2B3A55" />
              <Bar dataKey="expected" fill="#9F8772" />
            </BarChart>
          </ResponsiveContainer>
        </div>
      )}
      <Link to="/watcher" replace>
        <button className={styles.update}>Update Query</button>
      </Link>
    </div>
  )
}
