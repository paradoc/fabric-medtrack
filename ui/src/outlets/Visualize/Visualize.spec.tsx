import { render, waitFor, screen } from '@testing-library/react'
import React from 'react'
import { MemoryRouter } from 'react-router'
import { setupFetchStub } from '../../../setupTests'
import Visualize from './Visualize'

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useLocation: () => ({
    search: new URLSearchParams({
      g: 'test-generic',
      s: '2022-12-01',
      e: '2022-12-02',
    }),
  }),
}))

describe('Visualize', () => {
  window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
      disconnect: jest.fn(),
      observe: jest.fn(),
    }))

  beforeEach(() => {
    const fakeData = [
      {
        current: 1,
        expected: 1,
        generic_name: 'test-generic',
        total_dispatches: 1,
      },
    ]
    global.fetch = jest.fn().mockImplementation(setupFetchStub(fakeData))
  })

  afterEach(() => {
    // global.fetch.mockClear()
    // delete global.fetch
  })

  test('should match snapshot', async () => {
    const { container } = render(
      <MemoryRouter>
        <Visualize />
      </MemoryRouter>
    )
    await waitFor(() => {
      screen.getByTestId('chart')
    })
    expect(container).toMatchSnapshot()
  })
})
