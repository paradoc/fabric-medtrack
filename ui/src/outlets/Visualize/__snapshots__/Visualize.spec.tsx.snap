// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualize should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <header>
      Compliance rates for medicines:
       
       from 
      2022-12-01
       to 
      2022-12-02
    </header>
    <div
      data-testid="chart"
      style="height: 100%; width: 100%; flex: 1;"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 100%;"
      />
    </div>
  </div>
</div>
`;
